manual/manual/cmds/intf-c.etex

```diff
@@ -2702,3 +2698,67 @@ support libraries ("-lz") and the corresponding options
 ("-L/usr/local/zlib") must be given on all three invocations of "ocamlmklib",
 because they are needed at different times depending on whether shared
 libraries are supported.
+
+
+\section{s:c-internal-guidelines}{Disavowed topic: the internal runtime API}
+
+Let's start with a warning, there is \emph{no} stability guarantee if you are
+using internals of the OCaml runtime. If you really need access
+to the runtime internal API, this section provides some guidelines
+that may help you to write code that might not break on every new version
+of OCaml.
+
+\subsection{ss:c-internals}{CAML_INTERNALS}
+First, by default header files in the "caml/" subdirectory give only access
+to the public interface of the OCaml runtime.
+It is necessary to define the macro "CAML_INTERNALS" to get access to
+the lower-level interface. Once again, this lower-level interface is likely to
+change and break programs that use it.
+Another caveat is that this lower-level interface is only publicly accessible
+since OCaml 4.04.
+
+\subsection{ss:c-internal-variables}{Internal variables}
+Second, if you are using internal C variables, do not redefine them
+by hand. You should import those variables by including the corresponding
+header files. The representation of those variables has already changed once in
+OCaml 4.10, and is still under evolution.
+If your code relies on such internal and brittle properties, it will be broken
+at some point in time.
+
+For instance, rather than redefining "caml_young_limit":
+\begin{verbatim}
+extern int caml_young_limit;
+\end{verbatim}
+which breaks in OCaml $\ge$ 4.10, you should include the "minor_gc" header:
+\begin{verbatim}
+#include <caml/minor_gc.h>
+\end{verbatim}
+
+\subsection{ss:c-internal-macros}{OCaml version macros}
+Finally, if including the right headers is not enough, or if you need to support
+version older than OCaml 4.04, the header file "caml/version.h" should help
+you to define your own compatibility layer.
+This file provides macros defining the current ocaml major
+version "OCAML_VERSION_MAJOR", the minor version "OCAML_VERSION_MINOR", the
+patch level "OCAML_VERSION_PATCHLEVEL" and the full version
+"OCAML_VERSION" (which is the concatenation of the three previous macros).
+
+For example, if you need some specific handling for versions more recent than 4.10.0,
+you could write
+\begin{verbatim}
+#include <caml/version.h>
+#if OCAML_VERSION > 41000
+...
+#else
+...
+#endif
+\end{verbatim}
+
+The "caml/version.h" header also defines a macro for the compiler variant
+"OCAML_VERSION_ADDITIONAL" which is used to distinguish alpha and beta releases
+or release candidates.
+Finally, the version string is available as "OCAML_VERSION_STRING".
+
+\subsection{ss:c-internal-no}{A final word of caution}
+Do not access the domain state directly. \\
```

Comment:
This section needs to be expanded to document what domain state is.

Reply:
Does it? It is not clear to me that we have information on the domain state that would be useful to the narrow group of expert users diving into the runtime internals that are not compiler hackers. It might be easier to just remove the section.

Reply:
I posted the following comment earlier but it does not show up for me. I post it again:

I disagree with the advice to use `Caml_state_field`: it is only used if you want to be compatible with both short names and namespaced names. Ideally it is better if users do not use it at all. I'd rather advise people that they have to define CAML_NAME_SPACE and use `Caml_state->` directly.

However one should warn that it is internal, even if it can (currently) be accessed without CAML_INTERNALS. In the future, I even hope that it will be impossible to access Caml_state without it (#9179). Minimal documentation would count as partial fix for #9179

Reply:
As a simpler solution, I removed this section in the current version. We can add it later if needed.